### PR TITLE
Add EF.JsonExists(expression, path) DbFunction

### LINKED ISSUE #31136: Relational method for checking whether a path in a JSON exists
All relational databases support some way of checking whether a given JSON PATH exists in a JSON document. We should add a relational-level EF.Functions.JsonExists (or similar) which the different providers can translate to this.

Database   | Function
---------- | --------
PG         | [jsonb_path_exists](https://www.postgresql.org/docs/current/functions-json.html)
SQL Server | [JSON_PATH_EXISTS](https://learn.microsoft.com/en-us/sql/t-sql/functions/json-path-exists-transact-sql) (since SQL Server 2022)
SQLite     | [json_type](https://www.sqlite.org/json1.html#the_json_type_function) (returns NULL for non-existing path, as opposed to 'null' for JSON null values)
MariaDB    | [JSON_EXISTS](https://mariadb.com/kb/en/json_exists/)
MySQL      | [JSON_CONTAINS_PATH](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-contains-path)
Oracle     | [JSON_EXISTS](https://docs.oracle.com/en/database/oracle/oracle-database/12.2/adjsn/condition-JSON_EXISTS.html#GUID-8A0043D5-95F8-4918-9126-F86FB0E203F0)

### PR DESCRIPTION
**Pull request by mseada94**
- Fixes #31136

- [x] I've read the guidelines for [contributing](CONTRIBUTING.md) and seen the [walkthrough](https://youtu.be/9OMxy1wal1s?t=1869)
- [x] I've posted a comment on an issue with a detailed description of how I am planning to contribute and got approval from a member of the team
- [ ] The code builds and tests pass locally (also verified by our automated build checks)
- [x] Commit messages follow this format:
```
        Summary of the changes
        - Detail 1
        - Detail 2

        Fixes #bugnumber
```
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Code follows the same patterns and style as existing code in this repo

TODO
- [x] Define relational level `EF.JsonExists` DbFunction
- [x] Define SqlServer translation
- [x] Define Sqllite translation
- [ ] Add test for constant json expression translation success
- [ ] Add test for column (string) json expression translation success
- [ ] Add test for column (owned) json expression translation success
- [ ] Add test for column (string value converter) json expression translation success
- [ ] Add test for column (not of the previous types) json expression translation failure
- [ ] Add function test for 3 cases of the result (`true`, `false`, `null`)

### CODE REVIEW

File: src/EFCore.Sqlite.Core/Query/Internal/Translators/SqliteJsonFunctionsTranslator.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        SqlExpression? instance,
+        MethodInfo method,
+        IReadOnlyList<SqlExpression> arguments,
+        IDiagnosticsLogger<DbLoggerCategory.Query> logger)
+    {
+        if (JsonExistsMethodInfo.Equals(method)
+            && arguments[0].TypeMapping is SqliteJsonTypeMapping or StringTypeMapping)
+        {
+            // IIF(arguments_0 IS NULL, NULL, JSON_TYPE(arguments_0, arguments_1) IS NOT NULL)
+            return _sqlExpressionFactory.Function("IFF",
```
Comment (ranma42) [inline src/EFCore.Sqlite.Core/Query/Internal/Translators/SqliteJsonFunctionsTranslator.cs:50]: Any reason to avoid using a `Case` here?
It would automatically optimize out the `arguments_0 IS NULL` clause if it is false
Another minor thing is that a `Constant` might be analyzed better than a `Fragment`.

Also, shouldn't this check also verify that `arguments_1 IS NOT NULL`?
`IIF('{}' IS NULL, NULL, JSON_TYPE('{}', NULL) IS NOT NULL)` evaluates to `FALSE`, but IIUC the idea is that `JsonExists` should return `null` if either argument is `NULL`

Something like
```sql
CASE
  WHEN arguments_0 IS NOT NULL AND arguments_1 IS NOT NULL
    THEN JSON_TYPE(arguments_0, arguments_1) IS NOT NULL
END
```
?

Comment (mseada94) [reply src/EFCore.Sqlite.Core/Query/Internal/Translators/SqliteJsonFunctionsTranslator.cs:50]: > Any reason to avoid using a Case here?

There is actual reason, `IFF` function is equivalent to `CASE`, I use it for readability and simplicity reason only.
There is no compatibility issue with using `IFF`, It was introduced in SQLite 3.32.0 before `JSON_TYPE` which added on 3.38.0

> Another minor thing is that a Constant might be analyzed better than a Fragment.

I want to use it first but was not sure if it work, I did not find Null constant used on the rest of the code.
I want to try it after adding the test, please your advice on how to use it here.

---

> EDIT: I am not sure about the expected behavior for the 2nd argument

In SQL Server, the function will throw error if the second argument is null
```
Argument data type NULL is invalid for argument 2 of JSON_PATH_EXISTS function.
```
In SQLite, the function `json_type` when the second argument is null will return null

So behavior might vary between null or error, but generally second argument should not be null.
We could replace null path with empty string to throw the following error instead of returning null
```
SQLITE_ERROR: sqlite3 result code 1: bad JSON path: ''
```
or add condition to return custom error using `SQLITE_ERROR` function, However I don't think this will add value, we just need to keep it simple.

Comment (ranma42) [reply src/EFCore.Sqlite.Core/Query/Internal/Translators/SqliteJsonFunctionsTranslator.cs:50]: > > Any reason to avoid using a Case here?
> 
> There is actual no reason, `IFF` function is equivalent to `CASE`, I use it for readability and simplicity reason only. There is no compatibility issue with using `IFF`, It was introduced in SQLite 3.32.0 before `JSON_TYPE` which added on 3.38.0
> 

Ah, I see 👍 my recommendation was based on the fact that while Sqlite treats them in the same way, EFCore is actually able to simplify away a needless `Case` ;)

> > Another minor thing is that a Constant might be analyzed better than a Fragment.
> 
> I want to use it first but was not sure if it work, I did not find Null constant used on the rest of the code. I want to try it after adding the test, please your advice on how to use it here.

A null constant is indeed unusual, but it would be something like `_sqlExpressionFactory.Costant(null, method.ReturnType)`

For the whole expression (taking into account that we don't care about the nullability of the second argument) I would suggest something like:

```csharp
return _sqlExpressionFactory.Case(
    [new CaseWhenClause(
        _sqlExpressionFactory.IsNotNull(arguments[0]),
        _sqlExpressionFactory.IsNotNull(
            _sqlExpressionFactory.Function("JSON_TYPE",
                arguments,
                nullable: true,
                argumentsPropagateNullability: Statics.TrueArrays[2],
                returnType: typeof(string)))
    ],
    null);
```

Using this, the whole expression should simplify to `JSON_TYPE(arg0, arg1)` when the first argument is known to be non-nullable (and if it doesn't, I'd love to look into it, as it means some of the `CASE` optimizations are not working as intended 😇 ).

(The IFF expression would undergo some simplifications, too, but EFCore would get stuck with `IFF(FALSE, NULL, JSON_TYPE(arg0, arg1)` as it does not understand the semantics of `IFF`)

> 
> > EDIT: I am not sure about the expected behavior for the 2nd argument
> 
> In SQL Server, the function will throw error if the second argument is null
> 
> ```
> Argument data type NULL is invalid for argument 2 of JSON_PATH_EXISTS function.
> ```
> 
> In SQLite, the function `json_type` when the second argument is null will return null
> 
> So behavior might vary between null or error, but generally second argument should not be null. We could replace null path with empty string to throw the following error instead of returning null
> 
> ```
> SQLITE_ERROR: sqlite3 result code 1: bad JSON path: ''
> ```
> 
> or add condition to return custom error using `SQLITE_ERROR` function, However I don't think this will add value, we just need to keep it simple.

Agreed; there is probably little value in that and in that corner case it would be trivial for the user to handle `NULL`s with a conditional.

Comment (mseada94) [reply src/EFCore.Sqlite.Core/Query/Internal/Translators/SqliteJsonFunctionsTranslator.cs:50]: Thank you, I will push the modified expression.

**Comment by mseada94** [comment]
Sorry for taking to long with no updates, I faced some issues when writing the tests and I did not have the time to check them.

I fixed some of the translation code and used object datatype instead of generic which fail early and did not let the method translator to be called.

I had issue with SqlServer Compatibility Level, It was configured on the fixtures but the localdb on my device was not the latest version and it took me some time to figure it out.

There is some other issues:
1- The owned json column is failed on the translation before reaching the method translator on the argument visit check as the argument of type RelationalStructuralTypeShaperExpression, I am not sure how to handle this issue **please your advice**.
2- The Result asserting try to filter the data on memory to compare the results, while I did not add memory implementation for the method, I will check how the assertion is working and avoid memory translation if possible.

**Comment by roji** [comment]
@mseada94 no problem; when you feel the PR is ready for review, flag it as such and we'll take a look.

**Comment by mseada94** [comment]
@roji I was able to fix all issues, just the owned json property still not sure how to handler it and deal with the property as `ColumnExpression` instead of `RelationalStructuralTypeShaperExpression`

The owned property serialization and deserialization should be handled by the framework and it supports translating normal expressions so there is no need for using functions such as JsonExists nor JsonValue.

Should I ignore this case based on that?

**Comment by roji** [comment]
This was done in #37477.
